src/main/java/org/codehaus/mojo/exec/ExecJavaMojo.java

```diff
@@ -358,6 +371,123 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+                .map(i -> (Constructor<? extends Runnable>) i)
+                .filter(i -> Modifier.isPublic(i.getModifiers()))
+                .max(Comparator.<Constructor<? extends Runnable>, Integer>comparing(Constructor::getParameterCount))
+                .orElseThrow(() -> new IllegalArgumentException("No public constructor found for " + bootClass));
+        if (getLog().isDebugEnabled()) {
+            getLog().debug("Using constructor " + constructor);
+        }
+
+        Runnable runnable;
+        try { // todo: enhance that but since injection API is being defined at mvn4 level it is
+            // good enough
+            final Object[] args = Stream.of(constructor.getParameters())
+                    .map(param -> {
+                        try {
+                            return lookupParam(param);
+                        } catch (final ComponentLookupException e) {
+                            getLog().error(e.getMessage(), e);
+                            throw new IllegalStateException(e);
+                        }
+                    })
+                    .toArray(Object[]::new);
+            constructor.setAccessible(true);
+            runnable = constructor.newInstance(args);
+        } catch (final RuntimeException re) {
+            if (getLog().isDebugEnabled()) {
+                getLog().debug(
+                                "Can't inject " + runnableClass + "': " + re.getMessage() + ", will ignore injections",
+                                re);
+            }
+            final Constructor<? extends Runnable> declaredConstructor = runnableClass.getDeclaredConstructor();
+            declaredConstructor.setAccessible(true);
+            runnable = declaredConstructor.newInstance();
+        }
+        runnable.run();
+    }
+
+    private Object lookupParam(final java.lang.reflect.Parameter param) throws ComponentLookupException {
+        switch (param.getName()) {
+                // loose coupled to maven (wrapped with standard jvm types to not require it)
+            case "systemProperties": // Properties
+                return getSession().getSystemProperties();
+            case "systemPropertiesUpdater": // BiConsumer<String, String>
+                return propertiesUpdater(getSession().getSystemProperties());
+            case "userProperties": // Properties
+                return getSession().getUserProperties();
+            case "userPropertiesUpdater": // BiConsumer<String, String>
+                return propertiesUpdater(getSession().getUserProperties());
+            case "projectProperties": // Properties
+                return project.getProperties();
+            case "projectPropertiesUpdater": // BiConsumer<String, String>
+                return propertiesUpdater(project.getProperties());
+            case "highestVersionResolver": // Function<String, String>
+                return (Function<String, String>) ga -> {
+                    final int sep = ga.indexOf(':');
```

Comment:
This lambda can be defined in separate method, like in `propertiesUpdater`

Reply:
done